stan/math/rev/fun/diag_pre_multiply.hpp

```diff
@@ -0,0 +1,92 @@
+#ifndef STAN_MATH_REV_FUN_DIAG_PRE_MULTIPLY_HPP
+#define STAN_MATH_REV_FUN_DIAG_PRE_MULTIPLY_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/rev/core.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return the product of the diagonal matrix formed from the vector
+ * or row_vector and a matrix.
+ *
+ * @tparam T1 type of the vector/row_vector
+ * @tparam T2 type of the matrix
+ * @param m1 input vector/row_vector
+ * @param m2 input matrix
+ *
+ * @return product of the diagonal matrix formed from the
+ * vector or row_vector and a matrix.
+ */
+template <typename T1, typename T2, require_eigen_vector_t<T1>* = nullptr,
+          require_eigen_t<T2>* = nullptr,
+          require_any_rev_matrix_t<T1, T2>* = nullptr>
```

Comment:
There is some error in compiling with OpenCL. There should probably be an extra requires somewhere. @t4c1 can you look at this?

Reply:
That is a runtime error, not a compiling error. These tests are comparing OpenCL and CPU (this) implementation. Dimensions of matrix expressions somewhere in the new code do not match. Given what test fails, I guess this code does not handle matrices of size zero. If this is not caught by non-OpenCL tests a test case should be added.

Reply:
Thanks @t4c1, I'll look into it.

Reply:
Thanks @t4c1, makes sense

Reply:
The code Ben proposed above (exchanging replicate calls) fixed this issue. However, I am wondering why there was no error in the mix tests:
```
Eigen::MatrixXd a00(0, 0);
Eigen::VectorXd u0(0);
expect_diag_pre_multiply(u0, a00);
```

Reply:
Maybe the autodiff system is refusing to test a thing with zero inputs? It kinda doesn't make sense from that perspective. Maybe that code you quote is secretly doing nothing.

prim tests are responsible for testing edge case, values, and making sure errors throw in the right place. Mix makes sure if there are errors, that autodiff and non-autodiff versions of a function throw the same errors. So it makes sense to have duplicate tests in prim and mix for things. If the expect ad tests are refusing to test size zero things... I dunno.

Don't worry about it and I can look at this when I do a review when this is ready again.

Reply:
> Maybe the autodiff system is refusing to test a thing with zero inputs?

If so, that's a bug in the autodiff system (and probably my bad). The boundary condition should be that a nullary function (one with no args), is tested for value at all levels of autodiff.  In general, derivatives check derivatives of result vs. all arguments, and here there are no arguments, so the derivative checks automatically pass.

Reply:
OK, I'll add another test for that to prim, the autodiff test for zero inputs was already there so I also don't know if it does anything :) Thank you for all the information! As I said, your proposal fixed the issue with OpenCL test, so this shouldn't be a problem anymore.